File: internal/workspace/workspace.go
package workspace

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
	"unicode/utf8"

	"github.com/U109/api-subagents/internal/shared"
)

const maxFileBytes = 200000

var excluded = map[string]bool{".git": true, ".codex": true, ".ssh": true, ".aws": true, ".azure": true, ".pal": true, "node_modules": true, ".venv": true, "venv": true, "dist": true, "build": true, "coverage": true, ".tools": true, ".runtime": true, "release": true, ".desktop-resources": true}

var secretName = regexp.MustCompile(`(?i)^\.env($|\.)|\.(pem|key|p12|pfx)$|^(credentials|auth|models|runtime)\.json$`)

type Edit struct {
	OldText string `json:"oldText"`
	NewText string `json:"newText"`
}

type Proposal struct {
	Path        string  `json:"path"`
	OriginalSHA *string `json:"originalSha256"`
	Content     string  `json:"content"`
	Edits       []Edit  `json:"edits,omitempty"`
	Redacted    bool    `json:"redacted,omitempty"`
}

type Workspace struct {
	Root      string
	handle    *os.Root
	Proposals map[string]Proposal
	order     []string
}

type FileContent struct {
	Path    string `json:"path"`
	SHA     string `json:"sha256"`
	Content string `json:"content"`
}

// Blocked 排除路径各级的凭据、依赖和生成物，大小写变体也不能绕过。
func Blocked(path string) bool {
	for _, part := range strings.FieldsFunc(path, func(c rune) bool { return c == '/' || c == '\\' }) {
		if excluded[strings.ToLower(part)] || secretName.MatchString(part) {
			return true
		}
	}
	return false
}

// OpenWorkspace 固定真实根目录并打开受限文件句柄，底层禁止通过链接逃出项目。
func OpenWorkspace(root string) (*Workspace, error) {
	if !filepath.IsAbs(root) {
		return nil, errors.New("workspace 必须是绝对目录。")
	}
	actual, err := filepath.EvalSymlinks(root)
	if err != nil {
		return nil, errors.New("无法打开项目目录。")
	}
	handle, err := os.OpenRoot(actual)
	if err != nil {
		return nil, errors.New("workspace 必须是可访问的目录。")
	}
	return &Workspace{Root: actual, handle: handle, Proposals: map[string]Proposal{}}, nil
}

// Close 释放项目目录句柄，已保存的修改建议仍保留在内存中。
func (w *Workspace) Close() { _ = w.handle.Close() }

// Changes 按提出顺序返回修改建议，避免任务层依赖文件工具内部的排序结构。
func (w *Workspace) Changes() []Proposal {
	items := make([]Proposal, 0, len(w.order))
	for _, path := range w.order {
		items = append(items, w.Proposals[path])
	}
	return items
}

// Resolve 校验相对路径、排除规则及真实链接目标；新文件沿最近存在的父目录校验。
func (w *Workspace) Resolve(relative string, allowNew bool) (string, error) {
	if strings.ContainsAny(relative, "\x00:\\") || strings.HasPrefix(relative, "/") || filepath.IsAbs(relative) {
		return "", errors.New("请使用项目内相对路径，以 / 分隔。")
	}
	target := filepath.Join(w.Root, filepath.FromSlash(relative))
	rel, _ := filepath.Rel(w.Root, target)
	if !shared.Inside(w.Root, target) || Blocked(rel) {
		return "", errors.New("路径超出项目或属于排除目录/凭据文件。")
	}
	probe := target
	for {
		actual, err := filepath.EvalSymlinks(probe)
		if err == nil {
			actualRel, _ := filepath.Rel(w.Root, actual)
			if !shared.Inside(w.Root, actual) || Blocked(actualRel) {
				return "", errors.New("符号链接指向项目外或受限文件。")
			}
			break
		}
		if !os.IsNotExist(err) || !allowNew || probe == w.Root {
			return "", err
		}
		probe = filepath.Dir(probe)
	}
	return rel, nil
}

// Read 只读取最多 200 KB 的普通 UTF-8 文本，实际读取再次限额，防止文件中途增长。
func (w *Workspace) Read(relative string) (FileContent, error) {
	rel, err := w.Resolve(relative, false)
	if err != nil {
		return FileContent{}, err
	}
	f, err := w.handle.Open(rel)
	if err != nil {
		return FileContent{}, err
	}
	defer f.Close()
	stat, err := f.Stat()
	if err != nil || !stat.Mode().IsRegular() || stat.Size() > maxFileBytes {
		return FileContent{}, errors.New("只允许读取不超过 200 KB 的普通文本文件。")
	}
	data, err := io.ReadAll(io.LimitReader(f, maxFileBytes+1))
	if err != nil {
		return FileContent{}, err
	}
	if len(data) > maxFileBytes {
		return FileContent{}, errors.New("文件超过 200 KB。")
	}
	if bytes.ContainsRune(data, 0) || !utf8.Valid(data) {
		return FileContent{}, errors.New("不支持二进制或非 UTF-8 文件。")
	}
	return FileContent{relative, shared.Hash(data), string(data)}, nil
}

// List 逐批读取目录，最多扫描 1500 项并返回 500 项，跳过符号链接及受限名称。
func (w *Workspace) List(relative string, recursive bool) (shared.Object, error) {
	start, err := w.Resolve(relative, false)
	if err != nil {
		return nil, err
	}
	queue := []string{start}
	entries := []any{}
	scanned := 0
	for len(queue) > 0 && len(entries) < 500 && scanned < 1500 {
		rel := queue[0]
		queue = queue[1:]
		if _, err = w.Resolve(filepath.ToSlash(rel), false); err != nil {
			return nil, err
		}
		dir, err := w.handle.Open(rel)
		if err != nil {
			return nil, err
		}
		for {
			items, readErr := dir.ReadDir(32)
			for _, item := range items {
				scanned++
				path := filepath.Join(rel, item.Name())
				if !Blocked(path) && item.Type()&os.ModeSymlink == 0 {
					kind := "file"
					if item.IsDir() {
						kind = "directory"
						if recursive {
							queue = append(queue, path)
						}
					}
					entries = append(entries, shared.Object{"path": filepath.ToSlash(path), "type": kind})
				}
				if len(entries) >= 500 || scanned >= 1500 {
					break
				}
			}
			if readErr != nil || len(entries) >= 500 || scanned >= 1500 {
				break
			}
		}
		dir.Close()
		if !recursive {
			break
		}
	}
	return shared.Object{"entries": entries, "truncated": len(entries) >= 500 || scanned >= 1500}, nil
}

// Search 做字面量、忽略大小写的文本搜索，最多返回 80 处匹配及行号。
func (w *Workspace) Search(query, relative string) (shared.Object, error) {
	if query == "" || len([]rune(query)) > 300 {
		return nil, errors.New("搜索词须为 1–300 字符。")
	}
	listing, err := w.List(relative, true)
	if err != nil {
		return nil, err
	}
	matches := []any{}
	scanned := 0
	for _, v := range shared.Arr(listing["entries"]) {
		entry := shared.Obj(v)
		if entry["type"] != "file" {
			continue
		}
		file, err := w.Read(shared.Str(entry["path"]))
		if err != nil {
			continue
		}
		scanned++
		for index, line := range strings.Split(file.Content, "\n") {
			if strings.Contains(strings.ToLower(line), strings.ToLower(query)) {
				matches = append(matches, shared.Object{"path": file.Path, "line": index + 1, "text": shared.Clip(strings.TrimSuffix(line, "\r"), 400)})
				if len(matches) >= 80 {
					return shared.Object{"matches": matches, "truncated": true, "scannedFiles": scanned}, nil
				}
			}
		}
	}
	return shared.Object{"matches": matches, "truncated": listing["truncated"], "scannedFiles": scanned}, nil
}

// SameSHA 区分新文件的 null 指纹与已有文件的实际哈希。
func SameSHA(a, b *string) bool {
	if a == nil || b == nil {
		return a == nil && b == nil
	}
	return *a == *b
}

// Propose 暂存文件建议并核对原指纹，不直接修改项目；同路径的新建议替换旧建议。
func (w *Workspace) Propose(relative, content string, expected *string) (shared.Object, error) {
	if len(content) > maxFileBytes || !utf8.ValidString(content) {
		return nil, errors.New("建议必须是最多 200 KB 的 UTF-8 文本。")
	}
	rel, err := w.Resolve(relative, true)
	if err != nil {
		return nil, err
	}
	if rel == "." {
		return nil, errors.New("不能修改项目根目录。")
	}
	var actual *string
	current, err := w.Read(relative)
	if err == nil {
		actual = &current.SHA
	} else if !os.IsNotExist(err) {
		return nil, err
	}
	if !SameSHA(actual, expected) {
		return nil, errors.New("文件版本不匹配，请重新读取；新文件的 expectedSha256 应为 null。")
	}
	key := filepath.ToSlash(rel)
	if _, ok := w.Proposals[key]; !ok {
		if len(w.Proposals) >= 20 {
			return nil, errors.New("单个任务最多提出 20 个文件修改。")
		}
		w.order = append(w.order, key)
	}
	w.Proposals[key] = Proposal{Path: key, OriginalSHA: actual, Content: content}
	return shared.Object{"staged": true, "path": key, "applied": false}, nil
}

// ProposeEdit 使用唯一文本片段精确替换，避免为少量修改传递整份文件。
func (w *Workspace) ProposeEdit(relative string, edits []Edit, expected string) (shared.Object, error) {
	if len(edits) < 1 || len(edits) > 40 {
		return nil, errors.New("edits 必须包含 1–40 个精确替换片段。")
	}
	current, err := w.Read(relative)
	if err != nil {
		return nil, err
	}
	if current.SHA != expected {
		return nil, errors.New("文件版本不匹配，请重新读取。")
	}
	content := current.Content
	for _, edit := range edits {
		index := strings.Index(content, edit.OldText)
		if edit.OldText == "" || index < 0 || strings.Contains(content[index+1:], edit.OldText) {
			return nil, errors.New("oldText 必须精确且唯一匹配，请补充上下文。")
		}
		content = content[:index] + edit.NewText + content[index+len(edit.OldText):]
	}
	result, err := w.Propose(relative, content, &expected)
	if err == nil {
		proposal := w.Proposals[shared.Str(result["path"])]
		proposal.Edits = edits
		w.Proposals[proposal.Path] = proposal
	}
	return result, err
}

// Execute 只开放读取、搜索和暂存建议，不提供写盘、命令执行或其他 Agent 调用。
func (w *Workspace) Execute(name string, args any) (any, error) {
	if text, ok := args.(string); ok {
		if json.Unmarshal([]byte(text), &args) != nil {
			return nil, errors.New("工具参数 JSON 无效。")
		}
	}
	input, ok := args.(map[string]any)
	if !ok {
		return nil, errors.New("工具参数必须为对象。")
	}
	path, pathOK := input["path"].(string)
	if !pathOK {
		return nil, errors.New("path 必须是字符串。")
	}
	switch name {
	case "list_files":
		return w.List(path, input["recursive"] == true)
	case "read_file":
		return w.Read(path)
	case "search_text":
		return w.Search(shared.Str(input["query"]), path)
	case "propose_file":
		content, ok := input["content"].(string)
		if !ok {
			return nil, errors.New("content 必须是字符串。")
		}
		raw, exists := input["expectedSha256"]
		if !exists {
			return nil, errors.New("缺少 expectedSha256。")
		}
		var sha *string
		if raw != nil {
			text, ok := raw.(string)
			if !ok {
				return nil, errors.New("指纹类型无效。")
			}
			sha = &text
		}
		return w.Propose(path, content, sha)
	case "propose_edit":
		var edits []Edit
		if json.Unmarshal(shared.Marshal(input["edits"]), &edits) != nil {
			return nil, errors.New("替换片段无效。")
		}
		return w.ProposeEdit(path, edits, shared.Str(input["expectedSha256"]))
	}
	return nil, errors.New("未知文件工具。")
}

// Schema 声明严格对象参数，所有属性必填，禁止额外字段。
func Schema(properties shared.Object) shared.Object {
	required := []string{}
	for key := range properties {
		required = append(required, key)
	}
	return shared.Object{"type": "object", "properties": properties, "required": required, "additionalProperties": false}
}

var WorkspaceTools = []shared.Tool{
	{Name: "list_files", Description: "List bounded project files, excluding dependencies, credentials and generated directories.", Schema: Schema(shared.Object{"path": shared.Object{"type": "string"}, "recursive": shared.Object{"type": "boolean"}})},
	{Name: "read_file", Description: "Read a project UTF-8 file and SHA-256 using a relative path.", Schema: Schema(shared.Object{"path": shared.Object{"type": "string"}})},
	{Name: "search_text", Description: "Search literal text case-insensitively. Results include line numbers and truncation status.", Schema: Schema(shared.Object{"path": shared.Object{"type": "string"}, "query": shared.Object{"type": "string"}})},
	{Name: "propose_edit", Description: "Prefer small edits: unique oldText/newText replacements against original SHA. Submit all edits for a file together. Never writes files.", Schema: Schema(shared.Object{"path": shared.Object{"type": "string"}, "expectedSha256": shared.Object{"type": "string"}, "edits": shared.Object{"type": "array", "minItems": 1, "maxItems": 40, "items": Schema(shared.Object{"oldText": shared.Object{"type": "string"}, "newText": shared.Object{"type": "string"}})}})},
	{Name: "propose_file", Description: "Stage complete file content for parent review. Use original SHA, or null for a new file. Never writes files.", Schema: Schema(shared.Object{"path": shared.Object{"type": "string"}, "content": shared.Object{"type": "string"}, "expectedSha256": shared.Object{"type": []string{"string", "null"}}})},
}

type ApplyError struct {
	Cause   error
	Applied []string
}

// Error 保留部分成功的路径信息，由 CLI 另行输出 applied 数组。
func (e *ApplyError) Error() string { return e.Cause.Error() }

// ApplyProposals 由主 Agent 审查后执行；先核对所有文件，再逐个原子替换并报告部分成功。
func ApplyProposals(taskID, workspace string, paths []string, storage string) (shared.Object, error) {
	if !shared.IDPattern.MatchString(taskID) {
		return nil, errors.New("无效的 task_id。")
	}
	data, err := os.ReadFile(filepath.Join(storage, taskID+".json"))
	if err != nil {
		return nil, errors.New("无法读取任务记录。")
	}
	var task struct {
		Status    string     `json:"status"`
		Workspace string     `json:"workspace"`
		Changes   []Proposal `json:"changes"`
	}
	if json.Unmarshal(data, &task) != nil {
		return nil, errors.New("任务记录格式无效。")
	}
	w, err := OpenWorkspace(workspace)
	if err != nil {
		return nil, err
	}
	defer w.Close()
	if task.Status != "completed" || !shared.SamePath(w.Root, task.Workspace) {
		return nil, errors.New("只能应用已完成且属于指定项目的任务。")
	}
	selected := []Proposal{}
	for _, p := range task.Changes {
		if len(paths) == 0 || shared.Contains(paths, p.Path) {
			selected = append(selected, p)
		}
	}
	for _, path := range paths {
		found := false
		for _, p := range selected {
			if p.Path == path {
				found = true
			}
		}
		if !found {
			return nil, errors.New("选中文件不在任务建议中。")
		}
	}
	seen := map[string]bool{}
	for _, p := range selected {
		if p.Redacted {
			return nil, errors.New("含脱敏占位符的建议不能直接应用。")
		}
		result, err := w.Propose(p.Path, p.Content, p.OriginalSHA)
		if err != nil {
			return nil, err
		}
		key := shared.Str(result["path"])
		if runtime.GOOS == "windows" {
			key = strings.ToLower(key)
		}
		if seen[key] {
			return nil, errors.New("任务记录含重复文件建议。")
		}
		seen[key] = true
		if err = w.rejectLink(p.Path); err != nil {
			return nil, err
		}
	}
	applied := []string{}
	for _, p := range selected {
		if err = w.applyOne(p); err != nil {
			return nil, &ApplyError{err, applied}
		}
		applied = append(applied, p.Path)
	}
	return shared.Object{"task_id": taskID, "applied": applied, "changesApplied": true}, nil
}

// rejectLink 禁止替换符号链接本身，要求建议针对真实普通文件。
func (w *Workspace) rejectLink(path string) error {
	rel, err := w.Resolve(path, true)
	if err != nil {
		return err
	}
	info, err := w.handle.Lstat(rel)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return err
	}
	if info.Mode()&os.ModeSymlink != 0 {
		return errors.New("不能直接替换符号链接。")
	}
	return nil
}

// applyOne 使用受限目录句柄写入临时文件，提交前再次校验 SHA；新文件使用独占链接。
func (w *Workspace) applyOne(p Proposal) error {
	rel, err := w.Resolve(p.Path, true)
	if err != nil {
		return err
	}
	if err = w.handle.MkdirAll(filepath.Dir(rel), 0755); err != nil {
		return err
	}
	if _, err = w.Propose(p.Path, p.Content, p.OriginalSHA); err != nil {
		return err
	}
	mode := os.FileMode(0644)
	if p.OriginalSHA != nil {
		stat, err := w.handle.Stat(rel)
		if err != nil {
			return err
		}
		mode = stat.Mode().Perm()
	}
	temp := rel + ".proposal-" + shared.UUID() + ".tmp"
	defer w.handle.Remove(temp)
	f, err := w.handle.OpenFile(temp, os.O_WRONLY|os.O_CREATE|os.O_EXCL, mode)
	if err != nil {
		return err
	}
	_, err = f.WriteString(p.Content)
	if err == nil {
		err = f.Sync()
	}
	closeErr := f.Close()
	if err != nil {
		return err
	}
	if closeErr != nil {
		return closeErr
	}
	if _, err = w.Propose(p.Path, p.Content, p.OriginalSHA); err != nil {
		return err
	}
	if err = w.rejectLink(p.Path); err != nil {
		return err
	}
	if p.OriginalSHA == nil {
		return w.handle.Link(temp, rel)
	}
	return w.handle.Rename(temp, rel)
}

File: src/workspace.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import crypto from 'node:crypto';

const excluded = new Set([
  '.git',
  '.codex',
  '.ssh',
  '.aws',
  '.azure',
  '.pal',
  'node_modules',
  '.venv',
  'venv',
  'dist',
  'build',
  'coverage',
]);
const maxBytes = 200000;
/** 按路径各级名称排除依赖、生成物和常见凭据文件，避免被文件工具读出。 */
function blocked(rel) {
  return rel
    .split(/[\\/]/)
    .some(
      (p) =>
        excluded.has(p.toLowerCase()) ||
        /^\.env($|\.)|\.(pem|key|p12|pfx)$/i.test(p) ||
        /^(credentials|auth|models)\.json$/i.test(p),
    );
}
/** 用相对路径判断目标是否仍在项目内，不依赖容易误匹配的字符串前缀。 */
function inside(root, candidate) {
  const r = path.relative(root, candidate);
  return r === '' || (!r.startsWith('..' + path.sep) && r !== '..' && !path.isAbsolute(r));
}
/** 为读取的 UTF-8 文本生成版本指纹，应用建议前据此检测并发修改。 */
const hash = (text) => crypto.createHash('sha256').update(text).digest('hex');
export class Workspace {
  /** 保存真实项目根目录和内存建议表；建议不会直接影响磁盘内容。 */
  constructor(root) {
    this.root = root;
    this.proposals = new Map();
  }
  /** 要求绝对目录并解析真实路径，确保后续边界检查使用一致的项目根。 */
  static async open(root) {
    if (!path.isAbsolute(root)) throw new Error('workspace 必须是项目的绝对路径。');
    const actual = await fs.realpath(root);
    if (!(await fs.stat(actual)).isDirectory()) throw new Error('workspace 必须是目录。');
    return new Workspace(actual);
  }
  /** 校验相对路径、凭据规则与符号链接；新文件沿父目录检查，仍不得越过项目边界。 */
  async resolve(relative, allowNew = false) {
    if (
      typeof relative !== 'string' ||
      relative.includes('\0') ||
      relative.includes(':') ||
      path.isAbsolute(relative) ||
      relative.includes('\\')
    )
      throw new Error('请使用项目内相对路径，以 / 分隔。');
    const candidate = path.resolve(this.root, relative || '.');
    const normalized = path.relative(this.root, candidate);
    if (!inside(this.root, candidate) || blocked(normalized))
      throw new Error('路径超出项目或属于排除目录/凭据文件。');
    let probe = candidate;
    while (true) {
      try {
        const actual = await fs.realpath(probe);
        const actualRelative = path.relative(this.root, actual);
        if (!inside(this.root, actual) || blocked(actualRelative))
          throw new Error('符号链接指向项目外或受限文件。');
        break;
      } catch (e) {
        if (e.code !== 'ENOENT' || !allowNew || probe === this.root) throw e;
        probe = path.dirname(probe);
      }
    }
    return candidate;
  }
  /** 最多读取 200 KB 普通文本，同时返回 SHA；读取中再次限流，阻止 stat 后文件增长。 */
  async read(relative) {
    const file = await this.resolve(relative);
    const handle = await fs.open(file, 'r');
    try {
      const info = await handle.stat();
      if (!info.isFile() || info.size > maxBytes)
        throw new Error('只允许读取不超过 200 KB 的普通文本文件。');
      // stat 后文件仍可能增长；实际读取也必须有硬上限。
      const buffer = Buffer.alloc(maxBytes + 1);
      let size = 0;
      while (size < buffer.length) {
        const {bytesRead} = await handle.read(buffer, size, buffer.length - size, null);
        if (!bytesRead) break;
        size += bytesRead;
      }
      if (size > maxBytes) throw new Error('只允许读取不超过 200 KB 的普通文本文件。');
      const bytes = buffer.subarray(0, size);
      if (bytes.includes(0)) throw new Error('不支持二进制文件。');
      const text = bytes.toString('utf8');
      return {path: relative, sha256: hash(text), content: text};
    } finally {
      await handle.close();
    }
  }
  /** 有界遍历目录，最多返回 500 项并跳过符号链接；截断会在结果中明确标记。 */
  async list(relative = '.', recursive = false) {
    const start = await this.resolve(relative);
    const results = [];
    const queue = [start];
    let scanned = 0;
    while (queue.length && results.length < 500 && scanned < 1500) {
      const next = queue.shift();
      const relative = path.relative(this.root, next).split(path.sep).join('/');
      const dir = await this.resolve(relative);
      // 流式遍历，避免一次把大型目录的所有条目读入内存。
      for await (const item of await fs.opendir(dir)) {
        scanned++;
        const rel = path.relative(this.root, path.join(dir, item.name)).split(path.sep).join('/');
        if (!blocked(rel) && !item.isSymbolicLink()) {
          results.push({path: rel, type: item.isDirectory() ? 'directory' : 'file'});
          if (recursive && item.isDirectory()) queue.push(path.join(dir, item.name));
        }
        if (results.length >= 500 || scanned >= 1500) break;
      }
      if (!recursive) break;
    }
    return {entries: results, truncated: results.length >= 500 || scanned >= 1500};
  }
  /** 在允许读取的文件中做忽略大小写的字面搜索，最多返回 80 处匹配及行号。 */
  async search(query, relative = '.') {
    if (typeof query !== 'string' || !query || query.length > 300)
      throw new Error('搜索词须为 1–300 字符的普通文本。');
    const files = await this.list(relative, true);
    const matches = [];
    let count = 0;
    for (const entry of files.entries) {
      if (entry.type !== 'file') continue;
      try {
        const {content} = await this.read(entry.path);
        count++;
        const lines = content.split(/\r?\n/);
        for (let i = 0; i < lines.length; i++)
          if (lines[i].toLowerCase().includes(query.toLowerCase())) {
            matches.push({path: entry.path, line: i + 1, text: lines[i].slice(0, 400)});
            if (matches.length >= 80) return {matches, truncated: true, scannedFiles: count};
          }
      } catch {
        /* unreadable, binary and oversized files are omitted */
      }
    }
    return {matches, truncated: files.truncated, scannedFiles: count};
  }
  /** 校验版本后保存完整文件建议；新文件要求 SHA 为 null，同路径的新建议替换旧建议。 */
  async propose(relative, content, expectedSha256) {
    if (typeof content !== 'string' || Buffer.byteLength(content) > maxBytes)
      throw new Error('修改建议必须是最多 200 KB 的文本。');
    const target = await this.resolve(relative, true);
    if (target === this.root) throw new Error('不能修改项目根目录。');
    let current = null;
    try {
      current = await this.read(relative);
    } catch (e) {
      if (e.code !== 'ENOENT') throw e;
    }
    if ((current?.sha256 ?? null) !== expectedSha256)
      throw new Error('文件版本不匹配，请重新读取。新文件的 expectedSha256 应为 null。');
    const key = path.relative(this.root, target).split(path.sep).join('/');
    if (!this.proposals.has(key) && this.proposals.size >= 20)
      throw new Error('单个任务最多提出 20 个文件修改。');
    const proposal = {path: key, originalSha256: current?.sha256 ?? null, content};
    this.proposals.set(key, proposal);
    return {
      staged: true,
      path: key,
      applied: false,
      message: '建议已保存；项目文件未修改，主 Agent 负责检查和应用。',
    };
  }
  /**
   * 暂存小范围替换，避免模型为一行修改重写整份文件。
   * 每个 oldText 必须在当前工作副本中唯一出现；最终内容仍按原文件 SHA 校验，不直接写盘。
   */
  async proposeEdit(relative, edits, expectedSha256) {
    if (!Array.isArray(edits) || !edits.length || edits.length > 40)
      throw new Error('edits 必须包含 1–40 个精确替换片段。');
    const current = await this.read(relative);
    if (current.sha256 !== expectedSha256) throw new Error('文件版本不匹配，请重新读取。');
    let content = current.content;
    for (const edit of edits) {
      if (!edit || typeof edit.oldText !== 'string' || !edit.oldText || typeof edit.newText !== 'string')
        throw new Error('每个替换需要非空 oldText 和字符串 newText。');
      const index = content.indexOf(edit.oldText);
      if (index < 0 || content.indexOf(edit.oldText, index + 1) !== -1)
        throw new Error('oldText 必须精确且唯一匹配，请补充上下文后重试。');
      content = content.slice(0, index) + edit.newText + content.slice(index + edit.oldText.length);
    }
    const result = await this.propose(relative, content, expectedSha256);
    this.proposals.get(result.path).edits = structuredClone(edits);
    return result;
  }
  /** 调度模型可用的受限文件工具；拒绝未知工具和非对象参数，不提供命令执行能力。 */
  async execute(name, args) {
    if (!args || typeof args !== 'object' || Array.isArray(args))
      throw new Error('工具参数必须为对象。');
    if (name === 'list_files') return this.list(args.path ?? '.', Boolean(args.recursive));
    if (name === 'read_file') return this.read(args.path);
    if (name === 'search_text') return this.search(args.query, args.path ?? '.');
    if (name === 'propose_file') return this.propose(args.path, args.content, args.expectedSha256);
    if (name === 'propose_edit') return this.proposeEdit(args.path, args.edits, args.expectedSha256);
    throw new Error(`未知工具：${String(name).slice(0, 60)}`);
  }
}
/** 构造严格的工具参数对象定义，所有列出的属性均必填且不接受额外字段。 */
const schema = (properties) => ({
  type: 'object',
  properties,
  required: Object.keys(properties),
  additionalProperties: false,
});
export const workspaceTools = [
  {
    name: 'list_files',
    description:
      'List files inside the assigned project. Common dependency, credential and generated directories are excluded.',
    schema: schema({path: {type: 'string'}, recursive: {type: 'boolean'}}),
  },
  {
    name: 'read_file',
    description: 'Read a project text file and its SHA-256. Use a relative path with forward slashes.',
    schema: schema({path: {type: 'string'}}),
  },
  {
    name: 'search_text',
    description:
      'Search literal text, case-insensitively, in project files. Results are bounded; narrow the path when truncated.',
    schema: schema({path: {type: 'string'}, query: {type: 'string'}}),
  },
  {
    name: 'propose_edit',
    description:
      'Prefer for small edits: stage unique oldText/newText replacements against the original SHA. Submit all edits for a file together; a later proposal replaces the earlier one. Never writes files.',
    schema: schema({
      path: {type: 'string'},
      expectedSha256: {type: 'string'},
      edits: {
        type: 'array',
        minItems: 1,
        maxItems: 40,
        items: schema({oldText: {type: 'string'}, newText: {type: 'string'}}),
      },
    }),
  },
  {
    name: 'propose_file',
    description:
      'Stage complete proposed file content for the parent to inspect and apply. Does not modify any project file. Use SHA-256 from read_file, or null for a new file.',
    schema: schema({
      path: {type: 'string'},
      content: {type: 'string'},
      expectedSha256: {type: ['string', 'null']},
    }),
  },
];
